justfile

```diff
@@ -0,0 +1,148 @@
+# just has no idiom for setting a default value for an environment variable
+# so we shell out, as we need VIRTUAL_ENV in the justfile environment
+export VIRTUAL_ENV  := `echo ${VIRTUAL_ENV:-.venv}`
+
+export BIN := VIRTUAL_ENV + if os_family() == "unix" { "/bin" } else { "/Scripts" }
+export PIP := BIN + if os_family() == "unix" { "/python -m pip" } else { "/python.exe -m pip" }
+
+export DEFAULT_PYTHON := if os_family() == "unix" { "python3.8" } else { "python" }
+export IMAGE_NAME := "cohortextractor"
+
+
+
+# list available commands
+default:
+    @"{{ just_executable() }}" --list
+
+
+# clean up temporary files
+clean:
+    rm -rf .venv
+
+
+# ensure valid virtualenv
+virtualenv:
+    #!/usr/bin/env bash
+    set -euo pipefail
+
+    # allow users to specify python version in .env
+    PYTHON_VERSION=${PYTHON_VERSION:-$DEFAULT_PYTHON}
+
+    # create venv and upgrade pip
+    test -d $VIRTUAL_ENV || { $PYTHON_VERSION -m venv $VIRTUAL_ENV && $PIP install --upgrade pip; }
+
+    # ensure we have pip-tools so we can run pip-compile
+    test -e $BIN/pip-compile || $PIP install pip-tools
+
+
+_compile-dev src dst *args: virtualenv
+    #!/usr/bin/env bash
+    set -euo pipefail
+
+    # exit if src file is older than dst file (-nt = 'newer than', but we negate with || to avoid error exit code)
+    test "${FORCE:-}" = "true" -o {{ src }} -nt {{ dst }} || exit 0
+    $BIN/pip-compile --allow-unsafe --output-file={{ dst }} {{ src }} {{ args }}
+
+
+_compile-prod src dst *args: virtualenv
+    #!/usr/bin/env bash
+    set -euo pipefail
+
+    # exit if src file is older than dst file (-nt = 'newer than', but we negate with || to avoid error exit code)
+    test "${FORCE:-}" = "true" -o {{ src }} -nt {{ dst }} || exit 0
+    $BIN/pip-compile --allow-unsafe --extras=drivers --output-file={{ dst }} {{ src }} {{ args }}
+
+
+# update requirements.prod.txt if requirements.prod.in has changed
+requirements-prod *args:
+    "{{ just_executable() }}" _compile-prod setup.py requirements.prod.txt {{ args }}
+
+
+# update requirements.dev.txt if requirements.dev.in has changed
+requirements-dev *args: requirements-prod
+    "{{ just_executable() }}" _compile-dev requirements.dev.in requirements.dev.txt {{ args }}
```

Comment:
I think we could condense this like so:
```suggestion
_compile src dst env *args: virtualenv
    #!/usr/bin/env bash
    set -euo pipefail

    # exit if src file is older than dst file (-nt = 'newer than', but we negate with || to avoid error exit code)
    test "${FORCE:-}" = "true" -o {{ src }} -nt {{ dst }} || exit 0
    $BIN/pip-compile --allow-unsafe {{ if env == "dev" { "--extras=drivers" } else { "" } }} --output-file={{ dst }} {{ src }} {{ args }}


# update requirements.prod.txt if requirements.prod.in has changed
requirements-prod *args:
    "{{ just_executable() }}" _compile-prod setup.py requirements.prod.txt prod {{ args }}


# update requirements.dev.txt if requirements.dev.in has changed
requirements-dev *args: requirements-prod
    "{{ just_executable() }}" _compile-dev requirements.dev.in requirements.dev.txt dev {{ args }}
```

Reply:
Thanks! I wasn't sure what the idiomatic way of doing things is.

(Just's README is very long, and I didn't find it easy to find much in its documentation either.)

Reply:
tbh I'm still mostly figuring out what's possible in justfiles, more than what's idomatic or conventional. I found the [just programmer's manual](https://just.systems/man/en/chapter_1.html) hit the right spot for me in terms of being able to grok some useful features.

actually now you mention it, I think we can make this probably more idiomatic by removing the use of `just_executable()` & making `requirements-*` only have dependencies, like this:

```
# update requirements.dev.txt if requirements.dev.in has changed
requirements-dev *args: requirements-prod (_compile-dev "requirements.dev.in" "requirements.dev.txt" "dev" args)
```
(not sure if this is much better or not, but kind of nice not to invoke a child just process...)